android/app/src/main/java/com/audiobookshelf/app/player/MediaSessionCallback.kt

```diff
@@ -258,6 +277,80 @@ class MediaSessionCallback(var playerNotificationService:PlayerNotificationServi
   }
 
 
+
+  private fun debounceKeyEvent(keyEvent: KeyEvent?): Boolean {
+    // how does this work:
+    // - every keyDown and keyUp triggers a scheduled handler
+    // - another keyDown or keyUp cancels the scheduled handler and re-triggers it with new values
+    // - the handler takes clickCount:int and clickPressed:bool (if held down)
+    // - keyCodes increase the number of clicks (PlayPause+=1, Next+=2, Prev+=3)
+    // - depending on the number of clicks, the playerNotificationService handles the configured action
+    // problems:
+    // - the logs show pretty accurate click / hold detection, but it does not really translate well in the player
+    // - since the trigger is scheduled, it does run in a different thread
+    // - this leads to strange behaviour - probably easy to fix, but I'm no kotlin native (Coroutines)
+    // - probably after some actions the thread of the player is no longer accessible...
+    if (keyEvent?.action == KeyEvent.ACTION_UP) {
+      clickPressed = false
+      // Log.d(tag, "=== KeyEvent.ACTION_UP")
+
+    } else if (keyEvent?.action == KeyEvent.ACTION_DOWN) {
+      // Log.d(tag, "=== KeyEvent.ACTION_DOWN")
+
+      if(clickPressed) {
+        return false
+      }
+      clickPressed = true
+
+      when (keyEvent.keyCode) {
+        KeyEvent.KEYCODE_HEADSETHOOK,
+        KeyEvent.KEYCODE_MEDIA_PLAY,
+        KeyEvent.KEYCODE_MEDIA_PAUSE,
+        KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE -> {
+          clickCount++
+          Log.d(tag, "=== handleCallMediaButton: Headset Hook/Play/ Pause, clickCount=$clickCount")
+        }
+
+        KeyEvent.KEYCODE_MEDIA_NEXT -> {
+          clickCount += 2
+          Log.d(tag, "=== handleCallMediaButton: Media Next, clickCount=$clickCount")
+        }
+
+        KeyEvent.KEYCODE_MEDIA_PREVIOUS -> {
+          clickCount += 3
+          Log.d(tag, "=== handleCallMediaButton: Media Previous, clickCount=$clickCount")
+        }
```

Comment:
I'm not understanding the logic here. How do we know that KEYCODE_MEDIA_NEXT/PREVIOUS are clicks and not a separate next/prev button press?

Reply:
Android is handling specific shortcuts not emitting exact events, but overriding them by combined events:

- One click: Play/Pause
- Double click: Next
- Triple click: Previous
- Long press: Voice over (on newer Devices)

This cannot be prevented, because it is a system / kernel behaviour. 

So double-clicking does NOT emit 2 raw click events, but 1 single (automatically rewritten) event `KeyEvent.KEYCODE_MEDIA_NEXT`. 

I handle `KeyEvent.KEYCODE_MEDIA_NEXT` as if 2 single clicks had happened and increase the click count by 2. Same for the triple-click, which happens to be emitted as `KeyEvent.KEYCODE_MEDIA_PREVIOUS`  by Android instead of 3 single keydown/keyup event-combos.

<details>
This code wont work on Android Versions < 7.1. There is no possibility to detect `KEY_DOWN` events, because Android internally waits till `KEY_UP` doing nothing and then emits `KEY_DOWN` and `KEY_UP` together with no delay in between. So the scenario of holding down the key, that should emit something like:

- Press down: `KEY_DOWN`
- Wait: `// wait 1248ms`
- Release: `KEY_UP`

on Android 7.0 devices is 

- Press down: `// nothing`
- Wait: `// wait 1248ms`
- Release: `KEY_DOWN`, `KEY_UP`

</details>

Reply:
My understanding is that these events are emitted by the bluetooth device and so depending on the bluetooth device it might emit a KEYCODE_MEDIA_NEXT on 2 clicks if it is setup for that.
Or you might have a bluetooth device that has separate buttons next to play/pause that is for KEYCODE_MEDIA_NEXT.

Did you find any documentation on these KeyEvents that say when/how they are triggered?
I think this needs to be tested on a bluetooth device that has NEXT/PREV media buttons. I'll see if I have one. I think many bluetooth in vehicles have this.

Reply:
> My understanding is that these events are emitted by the bluetooth device and so depending on the bluetooth device it might emit a KEYCODE_MEDIA_NEXT on 2 clicks if it is setup for that.

Ah, now I understand. You mean a [device like this](https://www.amazon.de/exec/obidos/ASIN/B0CQCPVBGF/wwwgeschenke-inspiration-21) - a car bluetooth receiver with multiple buttons (e.g. for `next` and `prev`).

The PR is NON-DESTRUCTIVE, so that there are no behaviour changes, as long as the according setting (`extended headset controls`) is disabled, which it is by default. So no fear of breaking anything, that works fine at the moment.

I own such a car bluetooth recevier with buttons for `play/pause`, `next` and `previous` and tested it. It shows the following behaviour (which I would have expected):

- `play/pause`
  - single click = play/pause
  - double click = next
  - triple click = prev
  - 4 clicks or more (fast execution) = nothing 
  - multiple clicks (slow execution) = multiple play / pause events
- `next`
  - single click = next
  - 2 clicks or more (fast execution) = nothing (translated to 4 clicks or more)
  - multiple clicks (slow execution) = multiple next events
- `prev`
  - single click = prev
  - 2 clicks or more (fast execution) = nothing (translated to 6 clicks or more)
  - multiple clicks (slow execution) = multiple prev events

So I think it works like expected. Even better, on bluetooth devices with A SINGLE button, you can now use it for `next` and `prev`. 

The only unexpected behaviour might be that multiple fast `next` clicks won't do anything, as long as the timer is not hit - you have to click slower, but since this feature is disabled by default, I don't see this as a huge problem.